tls/s2n_send.c

```diff
@@ -127,37 +135,66 @@ ssize_t s2n_send(struct s2n_connection * conn, const void *buf, ssize_t size, s2
         conn->last_write_elapsed = elapsed;
     }
 
+    /* Find correct buf_idx and offset within it to start operations */
+    size_t user_data_consumed = offs + conn->current_user_data_consumed;
+    for (buf_idx = 0, buf_offs = 0; buf_idx < count; buf_idx++) {
+        if (user_data_consumed < bufs[buf_idx].iov_len) {
+            buf_offs = user_data_consumed;
+            break;
+        }
+        user_data_consumed -= bufs[buf_idx].iov_len;
+    }
+
     /* Now write the data we were asked to send this round */
     while (size - conn->current_user_data_consumed) {
-        struct s2n_blob in = {.data = ((uint8_t *)(uintptr_t) buf) + conn->current_user_data_consumed };
-        in.size = MIN(size - conn->current_user_data_consumed, max_payload_size);
-        /* If dynamic record size is enabled,
-         * use small TLS records that fit into a single TCP segment for the threshold bytes of data     
-         */
-        if (conn->active_application_bytes_consumed < (uint64_t) conn->dynamic_record_resize_threshold) {
-            int min_payload_size = s2n_record_min_write_payload_size(conn);
-            if (min_payload_size < in.size) {
-                in.size = min_payload_size; 
+        GUARD(s2n_stuffer_rewrite(&conn->out));
+
+        int space_remaining, space_needed;
+        do {
+            struct s2n_blob in = {.data = ((uint8_t *)(uintptr_t) bufs[buf_idx].iov_base) + buf_offs };
+            in.size = MIN(bufs[buf_idx].iov_len - buf_offs, max_payload_size);
+
+            /* If dynamic record size is enabled,
+             * use small TLS records that fit into a single TCP segment for the threshold bytes of data     
+             */
+            if (conn->active_application_bytes_consumed < (uint64_t) conn->dynamic_record_resize_threshold) {
+                int min_payload_size = s2n_record_min_write_payload_size(conn);
+                if (min_payload_size < in.size) {
+                    in.size = min_payload_size; 
+                }
             }
-        }
 
-        /* Don't split messages in server mode for interoperability with naive clients.
-         * Some clients may have expectations based on the amount of content in the first record.
-         */
-        if (conn->actual_protocol_version < S2N_TLS11 && writer->cipher_suite->record_alg->cipher->type == S2N_CBC && conn->mode != S2N_SERVER) {
-            if (in.size > 1 && cbcHackUsed == 0) {
-                in.size = 1;
-                cbcHackUsed = 1;
+            /* Don't split messages in server mode for interoperability with naive clients.
```

Comment:
Why would we ever split the I/O into multiple records? A caller could get this on their own by calling s2n_send() or the new s2n_sendv() with whatever boundaries they want anyway.

Reply:
So this code is exactly the same except that it has more indent because there is an outside loop. Are asking why I kept this piece of code?